larcoh/ancestry_plots.py

```diff
@@ -0,0 +1,337 @@
+"""
+Plot ancestry PCA analysis results
+"""
+
+import logging
+from collections import Counter
+from typing import List, Iterable, Optional
+
+import pandas as pd
+import numpy as np
+import hail as hl
+from bokeh.resources import CDN
+from bokeh.embed import file_html
+from bokeh.transform import factor_cmap, factor_mark
+from bokeh.plotting import ColumnDataSource, figure
+from bokeh.palettes import turbo  # flake: disable=F401
+from bokeh.models import CategoricalColorMapper, HoverTool
+from cpg_utils import Path
+from cpg_utils.hail_batch import reference_path, genome_build
+
+
+BG_LABEL = 'Provided ancestry (1KG+HGDP)'
+FG_LABEL = 'Inferred ancestry'
+
+
+def run(
+    out_path_pattern: Path,
+    sample_qc_ht_path: Path,
+    scores_ht_path: Path,
+    eigenvalues_ht_path: Path,
+    loadings_ht_path: Path,
+    inferred_pop_ht_path: Path,
+):
+    """
+    Generate plots in HTML format, write for each PC (of n_pcs) and
+    scope ("dataset", "population") plus for loadings) into
+    file paths defined by `out_path_pattern`.
+    """
+    sample_ht = hl.read_table(str(sample_qc_ht_path))
+    scores_ht = hl.read_table(str(scores_ht_path))
+    eigenvalues_ht = hl.read_table(str(eigenvalues_ht_path))
+    loadings_ht = hl.read_table(str(loadings_ht_path))
+    inferred_pop_ht = hl.read_table(str(inferred_pop_ht_path))
+
+    scores_ht = scores_ht.annotate(
+        pop=inferred_pop_ht[scores_ht.s].pop,
+        is_training=inferred_pop_ht[scores_ht.s].is_training,
+        dataset=sample_ht[scores_ht.s].dataset,
+    ).cache()
+
+    def key_by_external_id(ht_, meta_ht=None):
+        """
+        Assuming ht.s is a CPG id, replaces it with external ID,
+        assuming it's defined in meta_ht.external_id
+        """
+        if meta_ht is None:
+            meta_ht = ht_
+        ht_ = ht_.annotate(old_s=ht_.s).key_by('old_s')
+        ht_ = (
+            ht_.annotate(
+                s=hl.if_else(
+                    hl.is_defined(meta_ht[ht_.old_s]),
+                    meta_ht[ht_.old_s].external_id,
+                    ht_.old_s,
+                )
+            )
+            .key_by('s')
+            .drop('old_s')
+        )
+        return ht_
+
+    ht = key_by_external_id(scores_ht, sample_ht)
+
+    pop_full_names = hl.literal(
+        {
+            'nfe': 'Non-Finnish European',
+            'fin': 'Finnish',
+            'mid': 'Middle Eastern',
+            'oth': 'Other',
+            'afr': 'African/African American',
+            'eas': 'East Asian',
+            'sas': 'South Asian',
+            'amr': 'Latino/Admixed American',
+        }
+    )
+    ht = ht.annotate(pop=pop_full_names[ht.pop])
```

Comment:
I'm assuming the short 3-letter codes come from `gnomad.sample_qc.ancestry`? Could we import the full names from there too, to prevent going out of sync?